doc/specs/#1620 - Installer selector on multiple matches.md

```diff
@@ -0,0 +1,235 @@
+>
+```
+
+If a user enters the package name or ID then we should scan the whole list including the ones which are not shown to the user.
+
+If they enter a number then it should be in the range of the list shown to the user else it should be considered invalid input.
+
+### Refine the search
+
+A user can search results by entering text input. The text input should be compared with the Name and ID in the result.
+
+If a user enters a Name or ID and we found only an exact match then start installing the package else we should show the new refined list to the user.
+
+The text input is said to be invalid only if it doesn't match any of the names or IDs. The user will see an error for invalid input and will be asked to retry.
+
+If the user enters the name/ID and source or only source and there isn't any source then the user will see an error message which says that cannot find the package with the source.
+
+## UI/UX Design
+
+This is how the prompt should look like...
+
+```pwsh
+> winget install Example
+Multiple packages found matching.
+No. Name       Id                   Source
+-------------------------------------------
+ 1  Example    Example.Example1     msstore
+ 2  Example    Example.Example2     winget
+ ...
+ 20 Example20  Example.Example20    msstore
+There are 8 more packages.
+
+Please select a row number, or enter package name or ID to refine search...
+> _
+```
+
+If a user enters a row number between 1-20 then install the package
+
+```pwsh
+Please select a row number, or enter package name or ID to refine search...
+> 2
+Downloading Example (Example.Example2)...
+```
+
+If a user enters an invalid input.
+
+```pwsh
+Please select a row number, or enter package name or ID to refine search...
+> 23
+Please enter row number which are present in entries above.
+>
+```
+
+If the user enters a name or ID then we should refine the list.
+
+```pwsh
+...
+Please select a row number, or enter package name or ID to refine search...
+> Example
+
+Refined the search to Name/ID with `Example` in it.
+No. Name       Id                   Source
+-------------------------------------------
+ 1  Example    Example.Example1     msstore
+ 2  Example    Example.Example2     winget
+ 3  Example    Example.Example22    msstore
+
+Please select a row number, or enter package name or ID to refine search...
+> _
+```
+
+A user can use `--source` flag
+
+Now we only have a package with ID `Example.Example2` in the list because of the queries the user wrote.
+
+```pwsh
+Refined the search to Name/ID with `Example` in it.
+No. Name       Id                   Source
+-------------------------------------------
+ 1  Example    Example.Example1     msstore
+ 2  Example    Example.Example2     winget
+ 3  Example    Example.Example22    msstore
+
+Please select a row number, or enter package name or ID to refine search...
+> --source winget
+Downloading Example (Example.Example2)...
+```
+
+If the `--source` query matches nothing then ask the user to retry.
```

Comment:
I think this gets back to the idea of preference vs requirement. If the user specifies the source, it would be a requirement and not a preference, so if no packages are found using the source, then I would expect the result to be `No package found matching input criteria` and the menu to exit.

> A short note on preferences vs requirements. The original concept has always been that there are user preferences, and user requirements. User requirements must be met in order to proceed; example - a user sets a requirement of x64 packages only, an x86 package can never be installed. If a manifest only contains x86 versions, it won't be available to the user. User preferences change the prioritization, but do not need to be met to proceed; example - a user sets a preference of en-GB for Installer Locale. If a manifest contains the en-GB locale, it will be used, regardless of what the system locale is, but if the en-GB locale is not present, the installation will continue with the most applicable locale.
>  
> User requirements and preferences can be set through the settings file. However, any command line switches override the settings and are treated as a requirement, not as preference. So if the user has a requirement for x64 architecture, but specifically uses `--architecture x86`, then the requirement becomes x86 just for that run of the install command.

Reply:
Oh, that makes so much sense. Thanks for that. I always that it as a preference.